app/Http/Controllers/SearchController.php

```diff
@@ -72,51 +91,94 @@ public function search(Request $request)
 			$escaped_terms[] = SearchController::escape_like($term);
 		}
 
-		$id = Session::get('UserID');
+		$toplevel = $this->albumFunctions->getToplevelAlbums();
+		if ($toplevel === null) {
+			return Response::error('I could not find you.');
+		}
+
+		$albumIDs = [];
+		if ($toplevel['albums'] !== null) {
+			foreach ($toplevel['albums'] as $album) {
+				$albumIDs[] = $album->id;
+				if ($this->readAccessFunctions->album($album->id) === 1) {
+					$albumIDs = $this->albumFunctions->get_sub_albums($album, $albumIDs);
+				}
+			}
+		}
+		if ($toplevel['shared_albums'] !== null) {
+			foreach ($toplevel['shared_albums'] as $album) {
+				$albumIDs[] = $album->id;
+				if ($this->readAccessFunctions->album($album->id) === 1) {
+					$albumIDs = $this->albumFunctions->get_sub_albums($album, $albumIDs);
+				}
+			}
+		}
 
 		/**
-		 * Photos.
+		 * Albums.
 		 */
-		// for now we only look in OUR pictures
-		$query = Photo::where('owner_id', '=', $id);
+		$query = Album::whereIn('id', $albumIDs);
 		for ($i = 0; $i < count($escaped_terms); $i++) {
 			$escaped_term = $escaped_terms[$i];
 			$query = $query->Where(
-				function (Builder $query) use ($id, $escaped_term) {
+				function (Builder $query) use ($escaped_term) {
 					$query->where('title', 'like', '%'.$escaped_term.'%')
-						->orWhere('description', 'like', '%'.$escaped_term.'%')
-						->orWhere('tags', 'like', '%'.$escaped_term.'%');
+						->orWhere('description', 'like', '%'.$escaped_term.'%');
 				});
 		}
-		$photos = $query->get();
-
-		if ($photos != null) {
+		$albums = $query->get();
+		if ($albums != null) {
 			$i = 0;
-			foreach ($photos as $photo) {
-				$return['photos'][$i] = $photo->prepareData();
+			foreach ($albums as $album_model) {
+				$album = $album_model->prepareData();
+				if ($this->readAccessFunctions->album($album_model->id) === 1) {
+					$album['albums'] = $this->albumFunctions->get_albums($album_model);
+					$album = $album_model->gen_thumbs($album, $this->albumFunctions->get_sub_albums($album_model, [$album_model->id]));
+				}
+				$return['albums'][$i] = $album;
 				$i++;
 			}
 		}
 
-		/**
-		 * Albums.
+		/*
+		 * Photos.
+		 *
+		 * Begin by eliminating albums we can see but can't access (i.e.,
+		 * password-protected ones).
 		 */
-		$query = Album::where('owner_id', '=', $id);
+		for ($i = 0; $i < count($albumIDs);) {
```

Comment:
isn't this for loop useless ?
I mean it removes albumIDs from which we do not have access to the album.
However, unless I'm wrong **by construction**, they have this property. See line 103 104 111 and 112.

Reply:
This is a subtle point. Lines 103-104 and 111-112 ensure that we do not descend *into* an inaccessible album. However, the id of that inaccessible album _is_ included in `albumIDs` (see lines 102 and 110). That way we can match on the name of that inaccessible album (so, say, if every password-protected album has "with password" in its name, searching for "password" will return their complete list).

When searching for photos, however, which are inside the albums, we need to remove those inaccessible albums so that we don't leak the contents of password-protected albums.

Believe me, these code paths were tested and in fact this loop was added after the rest of the code, when I noticed during testing that the code was leaking photos.